README.md
ros_ethernet_rmp [![Build Status](https://api.travis-ci.org/WPI-RAIL/ros_ethernet_rmp.png)](https://travis-ci.org/WPI-RAIL/ros_ethernet_rmp)
================

#### ROS Wrapper for the Segway RMP Ethernet Python Driver
For full documentation, see [the ROS wiki](http://ros.org/wiki/ros_ethernet_rmp).

The code modified from examples provided by [SEGWAY Inc.](http://rmp.segway.com/). For full terms and conditions, see the [LICENSE](LICENSE) file.

This package serves primarily as a ROS wrapper for the [python_ethernet_rmp](https://github.com/WPI-RAIL/python_ethernet_rmp) package.

### License
For full terms and conditions, see the [LICENSE](LICENSE) file.

### Authors
See the [AUTHORS](AUTHORS.md) file for a full list of contributors.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ros_ethernet_rmp
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.8 (2015-02-10)
------------------
* Changed args in launch file so that they can be overriden
* Contributors: David Kent

0.0.7 (2015-01-19)
------------------
* Added approximation of caster wheel orientaiton
* Contributors: David Kent

0.0.6 (2014-09-02)
------------------
* fixed wheel direction for joint state publisher to work correctly with the fixed urdf
* Contributors: dekent

0.0.5 (2014-08-15)
------------------
* Merge pull request #17 from cmdunkers/develop
  updated yaw accel limit
* updated the yaw_accel_limit
* Merge branch 'develop' of https://github.com/cmdunkers/ros_ethernet_rmp into develop
* Merge pull request #2 from WPI-RAIL/develop
  cmdunkers
* Merge pull request #1 from WPI-RAIL/develop
  updates
* fixed joint state publisher issue
* Contributors: Chris Dunkers, Russell Toris, cmdunkers

0.0.4 (2014-08-05)
------------------
* fixed looking for wrong packages
* travis now pull messages from source
* moved messages to rmp_msgs
* Merge pull request #16 from cmdunkers/develop
  changed to rmp_msgs
* changed to rmp_msgs
* Battery monitor added to travis
* fixed CMakeLists bug
* Merge pull request #15 from cmdunkers/develop
  direction!
* direction!
* Merge pull request #14 from cmdunkers/develop
  radians!
* radians!
* missing thing
* launch file fix
* Merge pull request #12 from cmdunkers/develop
  Fixed issues with the wheel joint state
* fixed the modulo
* Fixed issues with the wheel joint state
  added the battery monitor to the launch file
* Contributors: Chris Dunkers, David Kent, Russell Toris, dekent

0.0.3 (2014-07-30)
------------------
* Merge pull request #11 from cmdunkers/develop
  added in joint state publisher from the segway feedback for the wheels
* added the correct joint lint names
* fixed package name
* added a joint state publisher for the wheels based ont he feedback
* Contributors: Chris Dunkers, Russell Toris

0.0.2 (2014-07-29)
------------------
* Merge pull request #10 from cmdunkers/develop
  fixed the yaw rates for carl
* fixed the yaw rates for carl
* Contributors: Chris Dunkers, Russell Toris

0.0.1 (2014-07-23)
------------------
* Merge pull request `#9 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/9>`_ from cmdunkers/develop
  Added a param to publish or not to publish the odometry transform for th...
* Added a param to publish or not to publish the odometry transform for the pose update
* Merge pull request `#8 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/8>`_ from cmdunkers/develop
  fixed rotation for the segway from RHR to +ve is clockwise
* fixed rotation for the segway from RHR to +ve is clockwise
* Merge pull request `#7 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/7>`_ from cmdunkers/develop
  made odom reference base_footprint
* changed the odom to be with respect to the base_footprint
* Merge pull request `#2 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/2>`_ from WPI-RAIL/develop
  renamed functions to underscores
* rename functions to underscores
* Merge pull request `#6 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/6>`_ from cmdunkers/develop
  Fixed the config parameters bug
* Fixed the issue of the user defined parameters not being set
* Found issue with config upload
* Merge pull request `#5 <https://github.com/WPI-RAIL/ros_ethernet_rmp/issues/5>`_ from cmdunkers/develop
  cleaned up the ethernet_rmp and pose updater files

scripts/rmp_joint_state.py
#!/usr/bin/env python

"""
Battery monitor for the Segway RMP platform.

Author:  Chris Dunkers, Worcester Polytechnic Institute
Version: June 23, 2014
"""
from rmp_msgs.msg import RMPFeedback
from python_ethernet_rmp.system_defines import *
from std_msgs.msg import String
from sensor_msgs.msg import JointState
import rospy
import math
import time
import os

class JointStateMonitor:

	def __init__(self):
		"""
		Initialize the subscriptions and publishers of the node.
		"""
		self.jointStatePub = rospy.Publisher('rmp_joint_states', JointState, queue_size = 'None')

		rospy.Subscriber("rmp_feedback", RMPFeedback, self.get_batt_state)
		
		"""
		Get the battery parameters
		"""
		self.has_two_wheels = rospy.get_param('~has_two_wheels',True)
		self.has_caster_wheel = rospy.get_param('~has_caster_wheel',True)
		self.link_left_front = rospy.get_param('~link_left_front','base_link_left_wheel_joint')
		self.link_right_front = rospy.get_param('~link_right_front','base_link_right_wheel_joint')
		self.link_left_rear = rospy.get_param('~link_left_rear','base_link_left_rear_wheel_joint')
		self.link_right_rear = rospy.get_param('~link_right_rear','base_link_right_rear_wheel_joint')
		self.link_caster = rospy.get_param('~link_caster','caster_plate_link_caster_link_joint')
		tire_diameter = rospy.get_param('/ethernet_rmp/my_tire_diameter',DEFAULT_TIRE_DIAMETER_M)
		self.circumference = math.pi*tire_diameter;
		self.prev_caster_pos = 0
		
	def get_batt_state(self, rmp):
		"""
		Read in the current RMP feedback and publish the pose
		:param rmp: the RMP feedback message
		"""
		rmp_items = rmp.sensor_items
		rmp_values = rmp.sensor_values
		
		joint_state = JointState()
		
		names = [self.link_left_front, self.link_right_front,
					self.link_left_rear, self.link_right_rear, self.link_caster]
		pos = [0,0,0,0,0]
		vel = [0,0,0,0,0]
		
		"""
		get the values for the feedback items needed
		"""
		for x in range(0, len(rmp_items)):
			if rmp_items[x] == 'left_front_pos_m':
				pos[0] = -((rmp_values[x]/self.circumference) % 1.0)*(2*math.pi)
			elif rmp_items[x] == 'right_front_pos_m':
				pos[1] = ((rmp_values[x]/self.circumference) % 1.0)*(2*math.pi)
			elif rmp_items[x] == 'left_rear_pos_m':
				pos[2] = -((rmp_values[x]/self.circumference) % 1.0)*(2*math.pi)
			elif rmp_items[x] == 'right_rear_pos_m':
				pos[3] = ((rmp_values[x]/self.circumference) % 1.0)*(2*math.pi)
			elif rmp_items[x] == 'left_front_vel_mps':
				vel[0] = -rmp_values[x]
			elif rmp_items[x] == 'right_front_vel_mps':
				vel[1] = rmp_values[x]
			elif rmp_items[x] == 'left_rear_vel_mps':
				vel[2] = -rmp_values[x]
			elif rmp_items[x] == 'right_rear_vel_mps':
				vel[3] = rmp_values[x]
				
		if self.has_two_wheels:
			num = 2
		else:
			num = 4
			
		for x in range(0,num):
			joint_state.name.append(names[x])
			joint_state.position.append(pos[x])
			joint_state.velocity.append(vel[x])
		
		"""
		calculate (very rough) caster wheel position
		"""
		if self.has_caster_wheel:
			left_vel = -vel[0]
			right_vel = vel[1]
			if abs(left_vel) < 0.01:
				left_vel = 0
			if abs(right_vel) < 0.01:
				right_vel = 0
			if left_vel != 0 or right_vel != 0:
				step = .04
				target = math.atan2(right_vel, left_vel) - math.pi/4
				if target > math.pi:
					target -= 2*math.pi
				if target < -math.pi:
					target += 2*math.pi
				pos[4] = self.prev_caster_pos
				#calculate direction to turn
				if abs(pos[4] - target) >= step:
					direction = 1
					if pos[4] > target:
						if abs(target - pos[4]) < (2*math.pi + target - pos[4]):
							direction = -1
					else:
						if (2*math.pi - (pos[4] - target)) < (pos[4] - target):
							direction = -1
					#turn to target
                       		 	pos[4] += direction*step
					while pos[4] <= -math.pi:
						pos[4] += 2*math.pi
					while pos[4] > math.pi:
						pos[4] -= 2*math.pi
				else:
					pos[4] = target
				self.prev_caster_pos = pos[4]
			else:
				pos[4] = self.prev_caster_pos

		joint_state.name.append(names[4])
		joint_state.position.append(pos[4])
		joint_state.velocity.append(vel[4])
		
		"""
		Publish the state of the wheels/joints
		"""
		joint_state.header.stamp = rospy.Time.now()
		self.jointStatePub.publish(joint_state)

if __name__ == "__main__":
	rospy.init_node("rmp_joint_state")
	jointState = JointStateMonitor()
	rospy.loginfo("RMP Joint State Started")
	rospy.spin()


scripts/rmp_exchange.py
#!/usr/bin/env python
import roslib 
roslib.load_manifest('ros_ethernet_rmp')
roslib.load_manifest('python_ethernet_rmp')
import rospy

"""--------------------------------------------------------------------
COPYRIGHT 2013 SEGWAY Inc.

Software License Agreement:

The software supplied herewith by Segway Inc. (the "Company") for its 
RMP Robotic Platforms is intended and supplied to you, the Company's 
customer, for use solely and exclusively with Segway products. The 
software is owned by the Company and/or its supplier, and is protected 
under applicable copyright laws.  All rights are reserved. Any use in 
violation of the foregoing restrictions may subject the user to criminal 
sanctions under applicable laws, as well as to civil liability for the 
breach of the terms and conditions of this license. The Company may 
immediately terminate this Agreement upon your use of the software with 
any products that are not Segway products.

The software was written using Python programming language.  Your use 
of the software is therefore subject to the terms and conditions of the 
OSI- approved open source license viewable at http://www.python.org/.  
You are solely responsible for ensuring your compliance with the Python 
open source license.

You shall indemnify, defend and hold the Company harmless from any claims, 
demands, liabilities or expenses, including reasonable attorneys fees, incurred 
by the Company as a result of any claim or proceeding against the Company 
arising out of or based upon: 

(i) The combination, operation or use of the software by you with any hardware, 
	products, programs or data not supplied or approved in writing by the Company, 
	if such claim or proceeding would have been avoided but for such combination, 
	operation or use.
 
(ii) The modification of the software by or on behalf of you 

(iii) Your use of the software.

 THIS SOFTWARE IS PROVIDED IN AN "AS IS" CONDITION. NO WARRANTIES,
 WHETHER EXPRESS, IMPLIED OR STATUTORY, INCLUDING, BUT NOT LIMITED
 TO, IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 PARTICULAR PURPOSE APPLY TO THIS SOFTWARE. THE COMPANY SHALL NOT,
 IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL OR
 CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER.
--------------------------------------------------------------------"""
from ros_ethernet_rmp.msg import rmpCommand, rmpFeedback
from python_ethernet_rmp.rmp_interface import RMP
from python_ethernet_rmp.system_defines import *
from python_ethernet_rmp.user_event_handlers import RMPEventHandlers
from python_ethernet_rmp.rmp_config_params import *
from geometry_msgs.msg import Twist
import sys,time,threading,Queue

"""
Define the update delay or update period in seconds. Must be greater
than the minimum of 0.01s
"""
UPDATE_DELAY_SEC = 0.02

"""
Define whether to log the output data in a file. This will create a unique CSV
file in ./RMP_DATA_LOGS with the filename containing a time/date stamp 
"""
LOG_DATA = True			 

"""
The platform address may be different than the one in your config
(rmp_config_params.py). This would be the case if you wanted to update 
ethernet configuration. If the ethernet configuration is updated the
system needs to be power cycled for it to take effect and this should
be changed to match the new values you defined in your config
"""
rmp_addr = ("192.168.0.40",8080) #this is the default value and matches the config

"""
Define the main function for the example. It creates a thread to run RMP and handles
passing the events to the user defined handlers in user_event_handlers.py
"""

class RMPExchange:
	def __init__(self):
		"""
		Initialze the thread 
		"""
				
		"""
		Read in the Ros Params and add them to a array to set the config params
		"""
		global UPDATE_DELAY_SEC, LOG_DATA, rmp_addr 
		update_time = rospy.get_param('/update_delay_sec',0.05)
		LOG_DATA = rospy.get_param('~log_data',False)
		ip_addr = rospy.get_param('~current_rmp_ip_addr',DEFAULT_IP_ADDRESS)
		port_num = rospy.get_param('~current_rmp_port_num',DEFAULT_PORT_NUMBER)
		self.isOmni = rospy.get_param('~is_omni ',False)
		try:
			dottedQuadToNum(ip_addr)
			if port_num > 0:
				rmp_addr = (ip_addr,port_num)
			else:
				rospy.logwarn("current_rmp_port_num is not a valid port number")	
		except:
			rospy.logwarn("current_rmp_ip_addr in not in dotted quad format")
		
		
		if update_time >= 0.01:
			UPDATE_DELAY_SEC = update_time
		else:
			rospy.logwarn("Update delay time is too fast, set to 0.01 seconds")
			UPDATE_DELAY_SEC = 0.01

		self.rmpParams = []
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_VELOCITY,rospy.get_param('~my_velocity_limit_mps',DEFAULT_MAXIMUM_VELOCITY_MPS)])
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_ACCELERATION,rospy.get_param('~my_accel_limit_mps2',DEFAULT_MAXIMUM_ACCELERATION_MPS2)])
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_DECELERATION,rospy.get_param('~my_decel_limit_mps2',DEFAULT_MAXIMUM_DECELERATION_MPS2)])
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_DTZ_DECEL_RATE,rospy.get_param('~my_dtz_rate_mps2',DEFAULT_MAXIMUM_DTZ_DECEL_RATE_MPS2)])
		self.rmpParams.append([RMP_CMD_SET_COASTDOWN_ACCEL,rospy.get_param('~my_coastdown_accel_mps2',DEFAULT_COASTDOWN_ACCEL_MPS2)])
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_TURN_RATE,rospy.get_param('~my_yaw_rate_limit_rps',DEFAULT_MAXIMUM_YAW_RATE_RPS)])
		self.rmpParams.append([RMP_CMD_SET_MAXIMUM_TURN_ACCEL,rospy.get_param('~my_yaw_accel_limit_rps2',DEFAULT_MAX_YAW_ACCEL_RPS2)])
		self.rmpParams.append([RMP_CMD_SET_TIRE_DIAMETER,rospy.get_param('~my_tire_diameter_m',DEFAULT_TIRE_DIAMETER_M)])
		self.rmpParams.append([RMP_CMD_SET_WHEEL_BASE_LENGTH,rospy.get_param('~my_wheel_base_length_m',DEFAULT_WHEEL_BASE_LENGTH_M)])
		self.rmpParams.append([RMP_CMD_SET_WHEEL_TRACK_WIDTH,rospy.get_param('~my_wheel_track_width_m',DEFAULT_WHEEL_TRACK_WIDTH_M)])
		self.rmpParams.append([RMP_CMD_SET_TRANSMISSION_RATIO,rospy.get_param('~my_gear_ratio',DEFAULT_TRANSMISSION_RATIO)])
		self.rmpParams.append([RMP_CMD_SET_INPUT_CONFIG_BITMAP,rospy.get_param('~my_config_bitmap',DEFAULT_CONFIG_BITMAP)])
		self.rmpParams.append([RMP_CMD_SET_ETH_IP_ADDRESS,rospy.get_param('~my_ip_address',DEFAULT_IP_ADDRESS)])
		self.rmpParams.append([RMP_CMD_SET_ETH_PORT_NUMBER,rospy.get_param('~my_port_num',DEFAULT_PORT_NUMBER)])
		self.rmpParams.append([RMP_CMD_SET_ETH_SUBNET_MASK,rospy.get_param('~my_subnet_mask',DEFAULT_SUBNET_MASK)])
		self.rmpParams.append([RMP_CMD_SET_ETH_GATEWAY,rospy.get_param('~my_gateway',DEFAULT_GATEWAY)])
		self.rmpParams.append([RMP_CMD_SET_USER_FB_1_BITMAP,rospy.get_param('~my_user_defined_feedback_bitmap_1',DEFAULT_USER_FB_1_BITMAP)])
		self.rmpParams.append([RMP_CMD_SET_USER_FB_2_BITMAP,rospy.get_param('~my_user_defined_feedback_bitmap_2',DEFAULT_USER_FB_1_BITMAP)])
		self.rmpParams.append([RMP_CMD_SET_USER_FB_3_BITMAP,rospy.get_param('~my_user_defined_feedback_bitmap_3',DEFAULT_USER_FB_1_BITMAP)])
		self.rmpParams.append([RMP_CMD_SET_USER_FB_4_BITMAP,rospy.get_param('~my_user_defined_feedback_bitmap_4',DEFAULT_USER_FB_1_BITMAP)])
						
		"""
		Create and response and command queue. The responses will be in the form of 
		a dictionary containing the vaiable name as the key and a converted value
		the names are defined in the feedback_X_bitmap_menu_items dictionaries if a particular
		variable is of interest
		"""
		self.rsp_queue = Queue.Queue()
		self.cmd_queue = Queue.Queue()
		self.in_flags  = Queue.Queue()
		self.out_flags = Queue.Queue()
		
		"""
		Create the thread to run RMP 
		"""
		self.my_thread = threading.Thread(target=RMP, args=(rmp_addr,self.rsp_queue,self.cmd_queue,self.in_flags,self.out_flags,UPDATE_DELAY_SEC,LOG_DATA))
		self.my_thread.daemon = True
		self.my_thread.start()
		
		"""
		Initialize my event handler class
		"""
		self.EventHandler = RMPEventHandlers(self.cmd_queue,self.rsp_queue,self.in_flags,self)
		
		"""
		Initialize the feedback publisher
		"""
		self.rmpFeedback = rmpFeedback()
		self.feedbackPub = rospy.Publisher('rmp_feedback', rmpFeedback, False, None, None)
		
	def __del__(self):
		"""
		send the signal to kill the thread
		"""
		self.in_flags.put(RMP_KILL)
		
		"""
		Wait for the thread to die
		"""
		while self.my_thread.isAlive():
			pass
		
		"""
		Exit main
		"""
		#sys.exit()
		print 'exited rmp_exchange'
				
	def sendMotionCommand(self,command):
		#TODO how to handle omni directional in cmd_vel
		if not self.isOmni:
			self.EventHandler.AddCommand([RMP_MOTION_CMD_ID,command.linear.x,command.angular.z])
		else:
			self.EventHandler.AddCommand([RMP_OMNI_MOTION_CMD_ID,command.linear.x,command.linear.y, command.angular.z])
	
	def sendRMPCommand(self,command):
		if command.cmd_id == 1280 or command.cmd_id == 1281: 
			cmd = [command.cmd_id, command.arg1, command.arg2] 
			if isValidCommand(cmd):
				self.EventHandler.AddCommand(cmd)
		elif command.cmd_id == 1536:
			cmd = [command.cmd_id, command.arg1, command.arg2, command.arg3] 
			if isValidCommand(cmd):
				self.EventHandler.AddCommand(cmd)
		
	def publishFeedback(self,fb_dict):
		snrValues = []
		snrItems = []
		fltValues = []
		fltItems = []
		ipValues = []
		ipItems = []
		
		"""
		get all of the values from the Feedback provided by the RMP
		"""
		for key, value in fb_dict.items():
			try:
				value = float(value)
				snrValues.append(value)
				snrItems.append(key)
			except:
				try:
					value = int(value,16)
					fltValues.append(value)
					fltItems.append(key)
				except:
					ipValues.append(value)
					ipItems.append(key)
					
		self.rmpFeedback.header.stamp = rospy.Time.now()
		self.rmpFeedback.header.frame_id = 'base_link'
		self.rmpFeedback.sensor_items = snrItems
		self.rmpFeedback.sensor_values = snrValues
		self.rmpFeedback.fault_status_items = fltItems
		self.rmpFeedback.fault_status_values = fltValues
		self.rmpFeedback.ip_info = ipItems
		self.rmpFeedback.ip_values = ipValues
		self.feedbackPub.publish(self.rmpFeedback)
			
	def isValidCommand(self, command):
		if command[0] == 1280 or command[0] == 1536:
			return True
		elif command[0] == 1281:
			if (command[1] >= 0 and command[1] <= 20) or (command[1] >=30 and  command[1] <=35):
				if command[1] == RMP_CMD_NONE:
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_VELOCITY:
					if command[2] > MAX_VELOCITY_MPS:
						rospy.logwarn("%s velocity was over max limit; set to %s", command[2], MAX_VELOCITY_MPS)
						command[2] = MAX_VELOCITY_MPS
					elif command[2] < MIN_VELOCITY_MPS:
						rospy.logwarn("%s velocity was under min limit; set to %s", command[2], MIN_VELOCITY_MPS)
						command[2] = MIN_VELOCITY_MPS
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_ACCELERATION:
					if command[2] > MAX_ACCELERATION_MPS2:
						rospy.logwarn("%s acceleration was over max limit; set to %s", command[2], MAX_ACCELERATION_MPS2)
						command[2] = MAX_ACCELERATION_MPS2
					elif command[2] < MIN_ACCELERATION_MPS2:
						rospy.logwarn("%s acceleration was under min limit; set to %s", command[2], MIN_ACCELERATION_MPS2)
						command[2] = MIN_ACCELERATION_MPS2
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_DECELERATION:
					if command[2] > MAX_DECELERATION_MPS2:
						rospy.logwarn("(%s) deceleration was over max limit; set to %s", command[2], MAX_DECELERATION_MPS2)
						command[2] = MAX_DECELERATION_MPS2
					elif command[2] < MIN_DECELERATION_MPS2:
						rospy.logwarn("(%s) deceleration was under min limit; set to %s", command[2], MIN_DECELERATION_MPS2)
						command[2] = MIN_DECELERATION_MPS2
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_DTZ_DECEL_RATE:
					if command[2] > MAX_DTZ_DECEL_RATE_MPS2:
						rospy.logwarn("(%s) dtz_decel was over max limit; set to %s", command[2], MAX_DTZ_DECEL_RATE_MPS2)
						command[2] = MAX_DTZ_DECEL_RATE_MPS2
					elif command[2] < MIN_DTZ_DECEL_RATE_MPS2:
						rospy.logwarn("(%s) dtz_decel was under min limit; set to %s", command[2], MIN_DTZ_DECEL_RATE_MPS2)
						command[2] = MIN_DTZ_DECEL_RATE_MPS2
					return True
				elif command[1] == RMP_CMD_SET_COASTDOWN_ACCEL:
					if command[2] > MAX_COASTDOWN_ACCEL_MPS2:
						rospy.logwarn("(%s) coastdown was over max limit; set to %s", command[2], MAX_COASTDOWN_ACCEL_MPS2)
						command[2] = MAX_COASTDOWN_ACCEL_MPS2
					elif command[2] < MIN_COASTDOWN_ACCEL_MPS2:
						rospy.logwarn("(%s) coastdown was under min limit; set to %s", command[2], MIN_COASTDOWN_ACCEL_MPS2)
						command[2] = MIN_COASTDOWN_ACCEL_MPS2
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_TURN_RATE:
					if command[2] > MAX_YAW_RATE_RPS:
						rospy.logwarn("%s yaw_rate was over max limit; set to %s", command[2], MAX_YAW_RATE_RPS)
						command[2] = MAX_YAW_RATE_RPS
					elif command[2] < MIN_YAW_RATE_RPS:
						rospy.logwarn("%s yaw_rate was under min limit; set to %s", command[2], MIN_YAW_RATE_RPS)
						command[2] = MIN_YAW_RATE_RPS
					return True
				elif command[1] == RMP_CMD_SET_MAXIMUM_TURN_ACCEL:
					if command[2] > MAX_YAW_ACCEL_RPS2:
						rospy.logwarn("%s yaw_accel was over max limit; set to %s", command[2], MAX_YAW_ACCEL_RPS2)
						command[2] = MAX_YAW_ACCEL_RPS2
					elif command[2] < MIN_YAW_ACCEL_RPS2:
						rospy.logwarn("%s yaw_accel was under min limit; set to %s", command[2], MIN_YAW_ACCEL_RPS2)
						command[2] = MIN_YAW_ACCEL_RPS2
					return True
				elif command[1] == RMP_CMD_SET_TIRE_DIAMETER:
					if command[2] > MAX_TIRE_DIAMETER_M:
						rospy.logwarn("%s tire diameter was over max limit; set to %s", command[2], MAX_TIRE_DIAMETER_M)
						command[2] = MAX_TIRE_DIAMETER_M
					elif command[2] < MIN_TIRE_DIAMETER_M:
						rospy.logwarn("%s tire diameter was under min limit; set to %s", command[2], MIN_TIRE_DIAMETER_M)
						command[2] = MIN_TIRE_DIAMETER_M
					return True
				elif command[1] == RMP_CMD_SET_WHEEL_BASE_LENGTH:
					if command[2] > MAX_WHEEL_BASE_LENGTH_M:
						rospy.logwarn("%s wheel base was over max limit; set to %s", command[2], MAX_WHEEL_BASE_LENGTH_M)
						command[2] = MAX_WHEEL_BASE_LENGTH_M
					elif command[2] < MIN_WHEEL_BASE_LENGTH_M:
						rospy.logwarn("%s wheel base was under min limit; set to %s", command[2], MIN_WHEEL_BASE_LENGTH_M)
						command[2] = MIN_WHEEL_BASE_LENGTH_M
					return True
				elif command[1] == RMP_CMD_SET_WHEEL_TRACK_WIDTH:
					if command[2] > MAX_WHEEL_TRACK_WIDTH_M:
						rospy.logwarn("%s wheel track was over max limit; set to %s", command[2], MAX_WHEEL_TRACK_WIDTH_M)
						command[2] = MAX_WHEEL_TRACK_WIDTH_M
					elif command[2] < MIN_WHEEL_TRACK_WIDTH_M:
						rospy.logwarn("%s wheel_track was under min limit; set to %s", command[2], MIN_WHEEL_TRACK_WIDTH_M)
						command[2] = MIN_WHEEL_TRACK_WIDTH_M
					return True
				elif command[1] == RMP_CMD_SET_TRANSMISSION_RATIO:
					if command[2] > MAX_TRANSMISSION_RATIO:
						rospy.logwarn("%s gear ratio was over max limit; set to %s", command[2], MAX_TRANSMISSION_RATIO)
						command[2] = MAX_TRANSMISSION_RATIO
					elif command[2] < MIN_TRANSMISSION_RATIO:
						rospy.logwarn("%s gear ratio was under min limit; set to %s", command[2], MIN_TRANSMISSION_RATIO)
						command[2] = MIN_TRANSMISSION_RATIO
					return True
				elif command[1] == RMP_CMD_SET_INPUT_CONFIG_BITMAP:
					if command[2] >= 0:
						return True
					else:
						rospy.logwarn("Invalid input config bitmap")
						return False
				elif command[1] == RMP_CMD_SET_ETH_IP_ADDRESS:
					try:
						dottedQuadToNum(command[2])
						return True
					except:
						rospy.logwarn("Invalid ip address")
						return False
				elif command[1] == RMP_CMD_SET_ETH_PORT_NUMBER:
					if command[2] > 0:
						return True
					else:
						rospy.logwarn("Invalid port number")
						return False
				elif command[1] == RMP_CMD_SET_ETH_SUBNET_MASK:
					try:
						dottedQuadToNum(command[2])
						return True
					except:
						rospy.logwarn("Invalid subnet mask")
						return False
				elif command[1] == RMP_CMD_SET_ETH_GATEWAY:
					try:
						dottedQuadToNum(command[2])
						return True
					except:
						rospy.logwarn("Invalid gateway")
						return False
				elif command[1] == RMP_CMD_SET_USER_FB_1_BITMAP:
					if command[2] >= 0:
						return True
					else:
						rospy.logwarn("Invalid user feedback 1 bitmap")
						return False
				elif command[1] == RMP_CMD_SET_USER_FB_2_BITMAP:
					if command[2] >= 0:
						return True
					else:
						rospy.logwarn("Invalid user feedback 2 bitmap")
						return False
				elif command[1] == RMP_CMD_SET_USER_FB_3_BITMAP:
					if command[2] >= 0:
						return True
					else:
						rospy.logwarn("Invalid user feedback 3 bitmap")
						return False
				elif command[1] == RMP_CMD_SET_USER_FB_4_BITMAP:
					if command[2] >= 0:
						return True
					else:
						rospy.logwarn("Invalid user feedback 4 bitmap")
						return False
				elif command[1] == RMP_CMD_SET_AUDIO_COMMAND:
					if command[2] >= 0 and command[2] <= 16:
						return True
					else:
						rospy.logwarn("Invalid audio command")
						return False
				elif command[1] == RMP_CMD_SET_OPERATIONAL_MODE:
					if command[2] >= 1 and command[2] <= 6:
						return True
					else:
						rospy.logwarn("Invalid operational mode")
						return False
				elif command[1] == RMP_CMD_SEND_SP_FAULTLOG: 
					if command[2] >= 0 and command[2] <= 1:
						return True
					else:
						rospy.logwarn("Invalid faultlog command")
						return False
				elif command[1] == RMP_CMD_RESET_INTEGRATORS: 
					if command[2] >= 0 or command[2] <= 31:
						return True
					else:
						rospy.logwarn("Invalid integrators reset command")
						return False
				elif command[1] == RMP_CMD_RESET_PARAMS_TO_DEFAULT:
					return True
				else:
					rospy.logwarn("Invalid command")
					return False
			else:
				rospy.logwarn("Invalid argument")
				return False
		else:
			rospy.logwarn("Invalid command id")
			return False
	
	def initRMPParams(self):
		for param in self.rmpParams:
			command = [RMP_CFG_CMD_ID, param[0], param[1]]
			if (True == self.isValidCommand(command)):
				self.EventHandler.AddCommand(command)
	
	def rmp_send_recv(self):
		"""
		Initialize the ROS node
		"""
		rospy.Subscriber("rmp_command", rmpCommand, self.sendRMPCommand)
		rospy.Subscriber("cmd_vel", Twist, self.sendMotionCommand)
		print "RMP exchange node started."
		
		"""
		Add a Feedback listener to the segway driver
		"""
		self.EventHandler.AddListener(self.publishFeedback)
		
		"""
		Initialize the parameters to the Segway
		"""
		self.initRMPParams()
				
		"""
		Generate a goto tractor event
		"""
		self.EventHandler.GotoTractor()
		print "In Tractor Mode"
		
		"""
		Main loop to continually empty yhe out_flags queue
		"""
		while not rospy.is_shutdown() and self.EventHandler._continue:
			while not self.out_flags.empty() and self.EventHandler._continue:
				self.EventHandler.handle_event[self.out_flags.get()]()
			
if __name__ == "__main__":
	rospy.init_node('rmp_exchange')
	rmp_command = RMPExchange()
	rmp_command.rmp_send_recv()	

